runtime/src/accounts_cache.rs

```diff
@@ -89,10 +105,35 @@ impl Deref for SlotCacheInner {
     }
 }
 
-#[derive(Debug, Clone)]
-pub struct CachedAccount {
+pub type CachedAccount = Arc<CachedAccountInner>;
+
+#[derive(Debug)]
+pub struct CachedAccountInner {
     pub account: AccountSharedData,
-    pub hash: Hash,
+    hash: RwLock<Option<Hash>>,
+    hash_slot: Slot,
```

Comment:
Why is it `hash_slot` and not just `slot` is there something about this slot that is specific to hashing? Or is it just the same slot that the account is stored at?

Reply:
I could call it 'slot'. My thinking was: I need the correct slot to compute the hash later. The slot is meant (by me) to be the same slot we would have used to calculate the hash had we done the hash calculation before adding the account to the cache. Thus, I called it 'hash_slot' in case there was confusion at some point later about why the slot is stored here.
I don't have enough system knowledge to know whether or when the slot could be different at a time when an account in the cache needs the slot for hashing or other potential purposes. Would you like me to rename it to 'slot'?

Reply:
`slot` is good. Just realized this data is kind of duplicated now. `cluster_type` never changes but now we store a copy for each cached account.

Reply:
> `cluster_type` never changes but now we store a copy for each cached account.

Yes, I do not like this. Do you know a good rust way to improve this?

Reply:
Well, there's standard plumbing it around from `AccountsDB::cluster_type`. The background service could have it's own copy which is move'ed in on initialization. I think the other places we need it already have `AccountsDB` reference?

Other than that there's `thread_local` or `lazy_static` but it's like global variables essentially and there is a perf hit for each access. Maybe it's a win overall from the original copy version as there would be less data polluting the caches. Would have to benchmark it to see.

Reply:
@jeffwashington @sakridge hi, rather late comment here. I think we can remove the cluster_type flag altogether because this ultimate code (https://github.com/solana-labs/solana/blob/b521f5027894f0ea4a07ac6dcc6c2419b7444868/runtime/src/accounts_db.rs#L2878-L2908) needing the flag is very old; we can remove the flag because we don't retain compatibility for the replayability since genesis. This will clean the code a lot across the call chain, assuming we don't ever change hash algorithm or its logic in near future ;)

related to this, I also want to clean this line of call-chain as well: https://github.com/solana-labs/solana/blob/b521f5027894f0ea4a07ac6dcc6c2419b7444868/runtime/src/accounts_db.rs#L3620-L3627
original pr: https://github.com/solana-labs/solana/pull/15763#discussion_r592080167


how about doing spring cleaning? :)

Reply:
yes, spring cleaning can occur.

Reply:
@ryoqun how do you want to get rid fo the cluster_type flag? Do we remove it from the [hash_stored_account function completely](https://github.com/solana-labs/solana/pull/16375)? is that what you're proposing?

Reply:
@jeffwashington yeah, that's correct. from time to time, we completely remove old dead-code behidn feature-gate like this: #15763